EOS-19534 Consul Hare Agent to be started by Hare in Mini-Provisioner Init Phase

SwapnilGaonkar7 opened the pull request:
Solution: Modified code to skip consul stop during shutdown, also note that consul will be restarted during bootstrap

knekrasov commented on utils/hare-bootstrap:
```diff
@@ -177,7 +178,7 @@ is_localhost() {
 
 TEMP=$(getopt --options hc: \
-              --longoptions help,mkfs,conf-dir:,debug \
+              --longoptions help,mkfs,skip_consul_start,conf-dir:,debug \
```
Optional: `s/skip_consul_start/skip-consul-start/`

knekrasov commented on utils/hare-shutdown:
```diff
@@ -55,13 +55,14 @@ def _get_processes_for_node() -> Dict[str, List[utils.Process]]:
         if svc != 'consul':
             utils.stop_parallel([proc for proc in processes[svc]])
-        else:
+        elif skip_consul_stop == False:
```
`eilf not skip_consul_stop:`

knekrasov replied:
Although I would prefer using 2 different [strategies](https://en.wikipedia.org/wiki/Strategy_pattern): one works as it used to, another one never stops consul.
The code of each strategy will be quite simple, the decision whether consul should be stopped or not is taken just once, at the very beginning. So lots of code will not need to propagate this `skip_consul_stop` flag and check it every now and then.

But again, if this is a draft and temporarily solution, please ignore this comment.

knekrasov commented on utils/hare-bootstrap:
```diff
@@ -209,8 +212,8 @@ $debug_p && set -x
 
 cdf=${1:-}
 
-if sudo systemctl --quiet is-active hare-consul-agent; then
```
Isn't it safer just create a copy of `hare-bootstrap` script and modify it - so the new one will be used by Mini Provisioner, but we'll not break the existing behavior for debugging and testing?

@mssawant what do you think?

mssawant replied:
And you mean we invoke the new one if `skip-consul-start` is specified to `hctl` interface? Although not for long but we will have to maintain multiple scripts and make sure that changes to one, if any, are incorporated in to other one too. On the other hand it will be cleaner to have a separate script.

knekrasov replied:
> And you mean we invoke the new one if skip-consul-start is specified to hctl interface? 

Not necessary. The script name will be different, so probably we'll not even need a flag.

mssawant replied:
I am thinking how would we invoke that different script, do you mean we just move to the new script and `hctl bootstrap` command invokes only the new script? or do you suggest to have a new `hctl command` altogether that invokes the new script?

knekrasov replied:
No. We can leave `hare-bootstrap` script as is so everybody who have been using `hctl bootstrap` will still get just the same behavior they would expect.

But we can copy `hare-bootstrap` -> `hare-provision` and put @SwapnilGaonkar7's changes there. Then it will be invoked like this: `hctl provision --mkfs /my/cdf/file.yaml`

The latter script will not start Consul.

Thoughts?

mssawant replied:
okay, so you are suggesting separate `hctl command`. I wonder if it would be confusing for the users of `hctl bootstrap`, we would need to document it well. In turn a flag like `skip-consul-start` seems more intuitive.
I think @SwapnilGaonkar7 is testing a "no change" option to hare-bootstrap (i.e. without `skip-consul-start`) May be we can just run `systemctl start hare-consul-agent` un-conditionally and handle the return code from systemd in case `hare-consul-agent` is already running.

mssawant commented on utils/hare-bootstrap:
```diff
@@ -40,6 +40,7 @@ Positional arguments:
 Options:
   --debug      Print commands and their arguments as they are executed.
   --mkfs       Execute m0mkfs.  *CAUTION* This wipes all Motr data!
+  --skip_consul_start   Consul will not be started and it is asuumed that it is already running
```
s/asuumed/assumed please

mssawant commented on utils/hare-bootstrap:
```diff
@@ -288,39 +293,41 @@ jq '[.[] | {key, value: (.value | @base64)}]' < $conf_dir/consul-kv.json |
-echo ' OK'
+    # Waiting for the Consul agents to get ready...
+    count=1
+    while (( $(get_ready_agents_nr) != $agents_nr )); do
```
Let's have this check unconditional to `skip_consul_start` value. Because even though Consul processes are running we need to make sure that all are running if not the we must error out.

mssawant commented on utils/hare-bootstrap:
```diff
@@ -288,39 +293,41 @@ jq '[.[] | {key, value: (.value | @base64)}]' < $conf_dir/consul-kv.json |
-    ssh $node "PATH=$PATH $(which mk-consul-env) \
-                          --mode server --bind $bind_ip --join $join_ip &&
-               sudo systemctl start hare-consul-agent" &
+if [ "$skip_consul_start" = false ] ; then
```
Now the question is if for some reason if the running Consul agent was stopped then who would restart it?
I would suggest the following,
- Continue iterating over all the nodes
- check if Consul is already running, if not start the Consul agent.

I think we also need to do this for node-start.

mssawant commented on utils/hare-shutdown:
```diff
@@ -73,16 +74,17 @@ def stop_process(processes: Dict[str, List[utils.Process]]):
-        pkill: str = 'sudo pkill --exact -KILL consul &> /dev/null'
-        utils.exec_custom(utils.ssh_prefix(leader_node) + pkill,
-                          show_err=False)
+    if skip_consul_stop == False:
```
As @knekrasov suggested, if we take out the consul shutdown code into a separate function, may be `stop_consul()` then we need not pass the flag to `shutdown_cluster()` in `main()` itself we can check `if skip_consul_stop` and invoke `stop_consul()` or not.

SwapnilGaonkar7 commented:
@mssawant @knekrasov I have modified code by just skipping consul stop during hctl shutdown, I have also addressed review comments which were valid after my changes.
Please have a look

knekrasov commented on utils/hare-shutdown:
```diff
@@ -55,13 +55,9 @@ def _get_processes_for_node() -> Dict[str, List[utils.Process]]:
 
 
-def stop_process(processes: Dict[str, List[utils.Process]]):
+def stop_consul(processes: Dict[str, List[utils.Process]], leader_node: str):
```
For me this name is misleading: we're stopping not only consul but all the processes.

`stop_all_with_consul()` maybe?

PS: that's why I like classes and overriden methods. Although the class name would be different (e.g. MiniProvFriendlyStopper and CommonStopper) but the main method would be named `stop()` in both cases 😀

SwapnilGaonkar7 replied:
Done

knekrasov reviewed (approved):
Minor comment only.

gowthamchinna commented:
run vm-deployment

gowthamchinna commented:
test vm-deployment

gowthamchinna commented:
test this please
